File: Data Structures & Algorithms/search-2d-matrix/submission-6.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        L, R = 0, len(matrix) - 1
        while L <= R:
            M = (L + R) // 2
            if target < matrix[M][0]:
                R = M - 1
            elif matrix[M][-1] < target:
                L = M + 1
            else:
                matrix = matrix[M]
                break
        
        if L > R:
            return False
        
        l, r = 0, len(matrix) - 1
        while l <= r:
            m = (l + r) // 2
            if target < matrix[m]:
                r = m - 1
            elif matrix[m] < target:
                l = m + 1
            else:
                return True
        
        return False

File: Data Structures & Algorithms/binary-search/submission-6.rs
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let (mut l, mut r) = (0i32, nums.len() as i32 - 1);
        while l <= r {
            let m = l + (r - l) / 2;
            if target > nums[m as usize] {
                l = m + 1;
            } else if target < nums[m as usize] {
                r = m - 1;
            } else {
                return m;
            }
        }
        -1
    }
}
